ouroboros-consensus/src/Ouroboros/Consensus/Util/OptNP.hs

```diff
@@ -209,3 +210,25 @@ combine = combineWith $ \case
     This1 x   -> x
     That1 y   -> y
     These1 {} -> error "combine: precondition violated"
+
+-- | Combines the first argument with the first element of the second argument.
+--
+-- Typically, the second argument will contain zero or one elements.
+--
+-- Precondition: there is no overlap between the first argument and the first
+-- element of the second argument: if there is a 'Just' at a given position in
+-- one, it must be 'Nothing' at the same position in the other.
+combineMaybeList ::
+     forall (f :: Type -> Type) xs.
+     -- 'These1' is not kind-polymorphic
+     HasCallStack
+  => Maybe (OptNP 'False f xs)
+  -> [OptNP 'False f xs]
+  -> [OptNP 'False f xs]
+combineMaybeList Nothing   yss      = yss
+combineMaybeList (Just xs) []       = [xs]
+combineMaybeList (Just xs) (ys:yss) = zipWith f xs ys : yss
```

Comment:
Too ad hoc, move it to `Cardano/Node.hs`, make it more specific and document what it is doing.